pkg/resource/node_resource_manager.go

```diff
@@ -202,36 +256,75 @@ func (o *NodeResourceManager) BuildNodeStatus(tsp *predictionapi.TimeSeriesPredi
-				if maxUsage < nextUsage {
-					maxUsage = nextUsage
+				if canNotBeReclaimedResource[resourceName] < nextUsage {
+					canNotBeReclaimedResource[resourceName] = nextUsage
 				}
 			}
-			var nextRecommendation float64
-			switch *resourceName {
-			case v1.ResourceCPU:
-				// cpu need to be scaled to m as ext resource cannot be decimal
-				nextRecommendation = (float64(node.Status.Allocatable.Cpu().Value()) - maxUsage) * 1000
-			case v1.ResourceMemory:
-				// unit of memory in prometheus is in Ki, need to be converted to byte
-				nextRecommendation = float64(node.Status.Allocatable.Memory().Value()) - (maxUsage * 1000)
-			default:
-				continue
-			}
-			if nextRecommendation < 0 {
-				klog.V(4).Infof("Unexpected recommendation,nodeName %s, maxUsage %v, nextRecommendation %v", node.Name, maxUsage, nextRecommendation)
-				continue
-			}
-			extResourceName := fmt.Sprintf(utils.ExtResourcePrefixFormat, string(*resourceName))
-			resValue, exists := node.Status.Capacity[v1.ResourceName(extResourceName)]
-			if exists && resValue.Value() != 0 &&
-				math.Abs(float64(resValue.Value())-
-					nextRecommendation)/float64(resValue.Value()) <= MinDeltaRatio {
-				continue
-			}
-			node.Status.Capacity[v1.ResourceName(extResourceName)] =
-				*resource.NewQuantity(int64(nextRecommendation), resource.DecimalSI)
-			node.Status.Allocatable[v1.ResourceName(extResourceName)] =
-				*resource.NewQuantity(int64(nextRecommendation), resource.DecimalSI)
 		}
 	}
+	return canNotBeReclaimedResource
+}
+
+func (o *NodeResourceManager) GetCanNotBeReclaimedResourceFromLocal() map[v1.ResourceName]float64 {
+	return map[v1.ResourceName]float64{
+		v1.ResourceCPU:    o.GetCpuCoreCanNotBeReclaimedFromLocal(),
+		v1.ResourceMemory: 0,
+	}
+}
+
+func (o *NodeResourceManager) GetCpuCoreCanNotBeReclaimedFromLocal() float64 {
+	if o.lastStateTime.Before(time.Now().Add(-20 * time.Second)) {
+		klog.V(1).Infof("NodeResourceManager local state has expired")
+		return 0
+	}
+
+	nodeCpuUsageTotalTimeSeries, ok := o.state[string(types.MetricNameCpuTotalUsage)]
+	if !ok {
+		klog.V(1).Infof("Can't get %s from NodeResourceManager local state", types.MetricNameCpuTotalUsage)
+		return 0
+	}
+	nodeCpuUsageTotal := nodeCpuUsageTotalTimeSeries[0].Samples[0].Value
+
+	var extResContainerCpuUsageTotal float64 = 0
+	extResContainerCpuUsageTotalTimeSeries, ok := o.state[string(types.MetricNameExtResContainerCpuTotalUsage)]
+	if ok {
+		extResContainerCpuUsageTotal = extResContainerCpuUsageTotalTimeSeries[0].Samples[0].Value
+	} else {
+		klog.V(1).Infof("Can't get %s from NodeResourceManager local state", types.MetricNameExtResContainerCpuTotalUsage)
+	}
+
+	var exclusiveCPUIdle float64 = 0
+	exclusiveCPUIdleTimeSeries, ok := o.state[string(types.MetricNameExclusiveCPUIdle)]
+	if ok {
+		exclusiveCPUIdle = exclusiveCPUIdleTimeSeries[0].Samples[0].Value
+	} else {
+		klog.V(1).Infof("Can't get %s from NodeResourceManager local state", types.MetricNameExclusiveCPUIdle)
+	}
+
```

Comment:
add some annotations for why use this formula?。。it may spend time to understand it

Reply:
ok